### PR TITLE
C++: Implement compilation_build_mode

### PR DESCRIPTION
**Pull request by calumgrant**
#### All query authors

- ~~[ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.~~
- ~~[ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.~~
- [x] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- ~~[ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.~~

#### Internal query authors only

- ~~[ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).~~
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- ~~[ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).~~

### CODE REVIEW

File: cpp/ql/lib/semmlecode.cpp.dbscheme [hunk 1]
```diff
@@ -46,6 +46,20 @@ compilation_args(
...
 
+/**
+ * Optionally, record the build mode for each compilation.
+ * The build mode should be the same for all compilations.
+ * Build mode:
+ *   0 = none
+ *   1 = manual
+ *   2 = auto
+ */
+#keyset[id, mode]
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply cpp/ql/lib/semmlecode.cpp.dbscheme:53]: Done. I did not realise this works in comments as well!

File: cpp/ql/lib/semmlecode.cpp.dbscheme.stats [hunk 2]
```diff
@@ -78,7 +78,7 @@
         </e>
         <e>
             <k>@using_enum_declaration</k>
-            <v>1</v>
+            <v>0</v>
```
Comment (jketema) [inline cpp/ql/lib/semmlecode.cpp.dbscheme.stats:81]: Looks like you didn't run the internal workflow we have to generate additional data from out tests.

Comment (calumgrant) [reply cpp/ql/lib/semmlecode.cpp.dbscheme.stats:81]: I ran the generate-db-stats workflow, downloaded the artefact, and merge was apparently successful:

```
[13:53:32] [DCA info] Normalizing stats by user-provided 'exprs'.
[13:53:32] [DCA info] Using stats from tests '~/Downloads/testsuite_stats.xml'.
[13:53:32] [DCA info] Using static stat files from /Users/calum/projects/codeql-dca/static-stats/cpp.
```

what am I missing?

Comment (jketema) [reply cpp/ql/lib/semmlecode.cpp.dbscheme.stats:81]: That's not possible for me to tell from this alone. Let's discuss internally?

File: cpp/downgrades/f0156f5f88ab5967c79162012c20f30600ca5ebf/upgrade.properties [hunk 3]
```diff
@@ -0,0 +1,3 @@
+description: Implement compilation_build_mode/2
+compatibility: backwards
```
Comment (jketema) [inline cpp/downgrades/f0156f5f88ab5967c79162012c20f30600ca5ebf/upgrade.properties:2]:
```suggestion
compatibility: full
```

**Review by jketema — APPROVED** [review]

**Pull request merged**